### PR TITLE
refactor(x/gamm): swap events

### PR DESCRIPTION
**Pull request by p0mvn**
Part of #1942

## Brief Changelog

Refactor and test gamm swap events. This does not change when the events are emitted. It only does some basic refactor to make events more testable.

Additionally, it unit tests and documents when swap events are emitted. All other events in gamm will be tested in incoming PRs

## Testing and Verifying

This change added tests for swap events at 3 levels of abstraction:
- internal events package
- keeper swap methods
- message server

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/gamm/keeper/multihop.go [hunk 1]
```diff
@@ -52,6 +52,9 @@ func (k Keeper) MultihopSwapExactAmountOut(
 	if err != nil {
 		return sdk.Int{}, err
 	}
+	if len(insExpected) == 0 {
+		return sdk.Int{}, nil
+	}
```
Comment (AlpinYukseloglu) [inline x/gamm/keeper/multihop.go:57]: Probably belongs in a separate PR but good addition (small enough to keep in this one imo)

File: x/gamm/keeper/msg_server_test.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	const (
+		tokenInMinAmount = 1
+		tokenIn          = 5
+	)
+
+	testcases := map[string]struct {
+		routes                []types.SwapAmountInRoute
+		tokenIn               sdk.Coin
+		tokenOutMinAmount     sdk.Int
+		expectError           bool
```
Comment (AlpinYukseloglu) [inline x/gamm/keeper/msg_server_test.go:29]: This should probably be flipped to `expectPass` to be in line with the rest of the table driven tests in our codebase. If there are any other `expectError`s, we should probably flip them to expectPass (or vice versa) for consistency (I'll make a separate issue about making this switch for other files)

File: x/gamm/spec/README.md [hunk 3]
```diff
@@ -968,6 +950,48 @@ osmosisd tx gamm swap-exact-amount-out 1000000ibc/27394FB092D2ECCD56123C74F36E4C
 ```
 
 ## Other resources
-* [Creating a liquidity bootstrapping pool](./client/docs/create-lbp-pool.md)
-* [Creating a pool with a pool file](./client/docs/create-pool.md)
 
+- [Creating a liquidity bootstrapping pool](./client/docs/create-lbp-pool.md)
+- [Creating a pool with a pool file](./client/docs/create-pool.md)
```
Comment (AlpinYukseloglu) [inline x/gamm/spec/README.md:955]: Looks like your IDE has the asterisk -> dash auto formatting too – should probably keep them as asterisks for now to be consistent with the rest of our markdown files but this is something I'm dealing with too

Comment (AlpinYukseloglu) [inline x/gamm/spec/README.md:955]:
```suggestion
* [Creating a liquidity bootstrapping pool](./client/docs/create-lbp-pool.md)
* [Creating a pool with a pool file](./client/docs/create-pool.md)
```

File: x/gamm/spec/README.md [hunk 4]
```diff
@@ -968,6 +950,48 @@ osmosisd tx gamm swap-exact-amount-out 1000000ibc/27394FB092D2ECCD56123C74F36E4C
...
+
+## Events
+
+There are 4 types of events that exist in GAMM:
+
+- `sdk.EventTypeMessage` - "message"
+- `types.TypeEvtPoolJoined` - "pool_joined"
+- `types.TypeEvtPoolExited` - "pool_exited"
+- `types.TypeEvtPoolCreated` - "pool_created"
+- `types.TypeEvtTokenSwapped` - "token_swapped"
```
Comment (AlpinYukseloglu) [inline x/gamm/spec/README.md:965]:
```suggestion
* `sdk.EventTypeMessage` - "message"
* `types.TypeEvtPoolJoined` - "pool_joined"
* `types.TypeEvtPoolExited` - "pool_exited"
* `types.TypeEvtPoolCreated` - "pool_created"
* `types.TypeEvtTokenSwapped` - "token_swapped"
```

File: x/gamm/spec/README.md [hunk 5]
```diff
@@ -968,6 +950,48 @@ osmosisd tx gamm swap-exact-amount-out 1000000ibc/27394FB092D2ECCD56123C74F36E4C
...
+- `sdk.AttributeKeyModule` - "module"
+  - The value is the module's name - "gamm".
+- sdk.AttributeKeySender
+  - The value is the address of the sender who created the swap message.
+- types.AttributeKeyPoolId
+  - The value is the pool id of the pool where swap occurs.
+- types.AttributeKeyTokensIn
+  - The value is the string representation of the tokens being swapped in.
+- types.AttributeKeyTokensOut
+  - The value is the string representation of the tokens being swapped out.
```
Comment (AlpinYukseloglu) [inline x/gamm/spec/README.md:997]:
```suggestion
* `sdk.AttributeKeyModule` - "module"
  * The value is the module's name - "gamm".
* sdk.AttributeKeySender
  * The value is the address of the sender who created the swap message.
* types.AttributeKeyPoolId
  * The value is the pool id of the pool where swap occurs.
* types.AttributeKeyTokensIn
  * The value is the string representation of the tokens being swapped in.
* types.AttributeKeyTokensOut
  * The value is the string representation of the tokens being swapped out.
```

**Review by AlpinYukseloglu — APPROVED** [review]
Great work! I left a few minor comments but overall LGTM

File: x/gamm/spec/README.md (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/spec/README.md:955]: I mean we can switch, I don't care that much. Just don't want a continual back and forth between the two.

File: x/gamm/keeper/msg_server_test.go (hunk 2 above)
Comment (p0mvn) [reply x/gamm/keeper/msg_server_test.go:29]: The best practice and a convention is to have `expectError error` where we can assert the actual kind of the error. Using a boolean is a shortcut to the original rule.

I'm not sure why we would have `expectPass` as a rule. I think it might make sense where the majority of test cases are errors and only one is valid. However, I think this can be decided on the case-by-case basis

File: x/gamm/keeper/internal/events/emit.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
+package events
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/gamm/keeper/internal/events/emit.go:1]: Having it in `types` allows it to be imported from any other package in the project. However, we only should be able to emit these events from the gamm keeper or message server. Therefore, making this package internal to the gamm `keeper` package allows for avoiding accidental imports anywhere else.

Another option could be is to simply define it as an "unexported" function in a separate file inside gamm keeper. With that, there are 2 issues:
- We would have to use `export_test.go` pattern to be able to export and test these event functions
- In my opinion, our keeper package is already bloated with functionality. `events` felt like a separate component that could be extracted

At the same time, I liked the ability to call these functions as `events.Emit...()` (by prefixing with `events.`. This makes it more visible than `k.emit...()` that events are being emitted. 

I'm planning to incrementally move all remaining events from the gamm keeper into this internal package and unit test them in PRs to come.

Let me know what you think

File: x/gamm/keeper/internal/events/emit.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+)
+
+func EmitSwapEvent(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, input sdk.Coins, output sdk.Coins) {
+	if ctx.EventManager() == nil {
+		return
+	}
```
Comment (mattverse) [inline x/gamm/keeper/internal/events/emit.go:14]: asking to learn: what are the supposable scenario that we don't ctx doesn't have EventManager? We didn't have this check before right?

File: x/gamm/types/events.go [hunk 8]
```diff
@@ -1,11 +1,5 @@
 package types
 
-import (
-	"strconv"
-
-	sdk "github.com/cosmos/cosmos-sdk/types"
-)
-
 const (
```
Comment (mattverse) [inline x/gamm/types/events.go:3]: Consider moving these and replacing these type calls with the newly added events emit. IMO could be confusing to maintain two different event packages for a single module.

**Review by mattverse — COMMENTED** [review]
Moving events into an internal package does seem like something unconventional, but I dont feel conflicted to this, as long as we figure out a way to maintain the two different packages for events, one in types, one in `events` package, which imo could be a confusing architecture

File: x/gamm/types/events.go (hunk 8 above)
Comment (p0mvn) [reply x/gamm/types/events.go:3]: The motivation behind leaving constants in types was to have them exposed / exported to clients who might want to use these constants to subscribe to events.

However, the emission logic should only be exposed to the gamm package, not the clients. That's why it was moved to an internal package inside gamm.

This is to preserve information hiding where we only export what clients should be able to see while unexport any package-private functionality.

Please let me know what you think

File: x/gamm/keeper/internal/events/emit.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply x/gamm/keeper/internal/events/emit.go:1]: In general, I prefer to be consistent with other modules. Typically, events are defined in `types/events.go`. 

Now I don't have a strong opinion here, but I will state that having events "public" can be a good and desirable thing. E.g. I've wrote many tools that exported and scraped on-chain data, where I used events to query for txs/blocks. Being able to directly reference the events was immensely useful instead of having to redefine them.  

So for this reason, I say just keep them in `types/`.

File: x/gamm/keeper/internal/events/emit.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
+package events
+
+import (
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+)
+
```
Comment (alexanderbez) [inline x/gamm/keeper/internal/events/emit.go:10]: Unless we have multiple places where we call `Emit*Event`, I'd say extracting these into functions is overkill IMO.

File: x/gamm/keeper/internal/events/emit.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattverse) [reply x/gamm/keeper/internal/events/emit.go:1]: agreed here!

File: x/gamm/types/events.go (hunk 8 above)
Comment (p0mvn) [reply x/gamm/types/events.go:3]: Let's continue discussing this in this thread please:
https://github.com/osmosis-labs/osmosis/pull/1983#discussion_r915894294

File: x/gamm/keeper/internal/events/emit.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/gamm/keeper/internal/events/emit.go:1]: Thanks for the feedback.

I agree about the need for having event names public by clients/tools, For this reason, they are still kept in types as per suggestion:
https://github.com/osmosis-labs/osmosis/blob/ad633616412c32d8918411e06120b65a9fd84a07/x/gamm/types/events.go#L3-L14

This is also consistent with other modules such as `epochs`:
https://github.com/osmosis-labs/osmosis/blob/716c6e4e0cb01bc089559bc3914fa17c93f73f8b/x/epochs/types/events.go#L1-L9

Contrary to event names, clients should never be able to emit the events. This is something that only the gamm module and internal logic should be able to do. That's why this logic was moved to an internal `events` package that is never exposed to outside packages or tools.

Having it in `types` would be information leakage because we are exposing functionality that is not meant to be used by any other module, and, as a result, opening up a vector to make a mistake by accidentally emitting events from somewhere else.

@alexanderbez @mattverse @ValarDragon  please let me know what you think

File: x/gamm/keeper/internal/events/emit.go (hunk 9 above)
Comment (p0mvn) [reply x/gamm/keeper/internal/events/emit.go:10]: There are indeed multiple places where events are emitted. In addition, the events differ as well as their attributes.

Having each event encapsulated in a function allows for asserting the kind of attributes emitted by events in tests. Otherwise, they would be left completely untested

An example where attributes are tested:
https://github.com/osmosis-labs/osmosis/blob/ad633616412c32d8918411e06120b65a9fd84a07/x/gamm/keeper/internal/events/emit_test.go#L61-L70

**Review by ValarDragon — APPROVED** [review]
LGTM

I do share the hesitance around the `internal/events` package, versus making these methods public.

Not sure that I have a particularly thought through opinion here, versus just general hesitance perhaps due to unfamiliarity. The API difference here is in the concrete, just making the old keeper-only private methods (that I liked) now moved to an internal package.
